mono/metadata/class.c

```diff
@@ -5494,6 +5494,10 @@ mono_class_has_parent_and_ignore_generics (MonoClass *klass, MonoClass *parent)
 		if (parent == mono_class_get_generic_type_definition (m_class_get_supertypes (klass) [i]))
 			return TRUE;
 	}
+
+	if (mono_class_implements_interface (klass, parent))
```

Comment:
1. I think this needs a comment.
2. Would we ever return `TRUE` here when `parent` isn't an interface?
3. Do we need to ignore generic instantiations here too?  I'm thinking of some variant of the example in #13550 but with generics:
   ```csharp
   public interface I1<X>
   {
       protected interface T1
       {
           void M1();
       }
   
       protected class T2
       {}
   
       protected struct T3
       {}
   
       protected enum T4
       {
           B
       }
   }
   
   class Test1 : I1<int>
   {
       static void Main()
       {
           I1<object>.T1 a = new Test2();
           a.M1();
           System.Console.WriteLine(new I1<object>.T2());
           System.Console.WriteLine(new I1<object>.T3());
           System.Console.WriteLine(I1<objec>.T4.B);
       }
   
       class Test2 : I1<object>.T1
       {
           public void M1()
           {
               System.Console.WriteLine("M1");
           }
       }
   }
   ```
   (This might not be the exactly right example - I haven't tried to compile or run it; but the idea is that the accessing class Test1 implements `I1<int>` while it is trying to access a member from `I1<object>` which it does not implement, so it's not assignable.)

Reply:
1) Okay
2) I don't know if I understand it right, but the plan is to return TRUE when the class implements the interface, so the parent should be an interface, I've changed the code.
3) Using CoreClr this example works and the result is this:
M1
I1`1+T2[System.Object]
I1`1+T3[System.Object]
B
Using mono it wasn't working but I fixed this in the new Commit. This new commit also fixes #13551 .